core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -570,28 +577,49 @@ private[spark] class BlockManager(
     preferredLocs ++ otherLocs
   }
 
-  private def doGetRemote(blockId: BlockId, asBlockResult: Boolean): Option[Any] = {
+  /**
+   * VisibleForTesting
+   */
+  private[spark] def doGetRemote(blockId: BlockId, blockTransferService: BlockTransferService,
+                                 asBlockResult: Boolean): Option[Any] = {
     require(blockId != null, "BlockId is null")
+    var runningFailureCount = 0
+    var totalFailureCount = 0
     val locations = getLocations(blockId)
-    var numFetchFailures = 0
-    for (loc <- locations) {
+    val maxFetchFailures = locations.size
+    var locationIterator = locations.iterator
+    while (locationIterator.hasNext) {
```

Comment:
just a thought: would this be easier to reason about as a tail-recursive method?  would eliminate the need for the vars and returns.